migration-0.70/main.go

```diff
@@ -0,0 +1,870 @@
+
+			} else if _, ok := allScopeAttrs[clean]; ok {
+				scopeAttrs[clean] = v
+
+			} else if _, ok := allPointAttrs[clean]; ok {
+				pointAttrs[clean] = v
+			}
+		}
+
+		resFP := internal.NewFingerprint(internal.ResourceFingerprintType,
+			internal.InitialOffset, attrsToPMap(resourceAttrs), nil)
+		scopeFP := internal.NewFingerprint(internal.ScopeFingerprintType,
+			resFP.Hash(), attrsToPMap(scopeAttrs), nil)
+
+		temporality := toOtelTemporality(norm.Temporality)
+
+		pointFP := internal.NewFingerprint(
+			internal.PointFingerprintType,
+			scopeFP.Hash(),
+			attrsToPMap(pointAttrs),
+			map[string]string{"__temporality__": temporality.String()},
+		)
+
+		var unNorm ts
+		unNorm.Env = norm.Env
+		unNorm.Temporality = norm.Temporality
+		unNorm.MetricName = md.unNormMetricName
+		unNorm.Description = norm.Description
+		unNorm.Unit = norm.Unit
+		unNorm.Type = norm.Type
+		unNorm.IsMonotonic = norm.IsMonotonic
+		unNorm.Fingerprint = pointFP.HashWithName(unNorm.MetricName)
+		unNorm.UnixMilli = norm.UnixMilli
+		unNorm.Normalized = false
+		unNorm.ScopeAttrs = scopeFP.AttributesAsMap()
+		unNorm.ResourceAttrs = resFP.AttributesAsMap()
+		unNorm.Attrs = pointFP.AttributesAsMap()
+		unNorm.Labels = internal.NewLabelsAsJSONString(
+			unNorm.MetricName, unNorm.Attrs,
+			unNorm.ScopeAttrs, unNorm.ResourceAttrs)
+
+		fingerprintMap[norm.Fingerprint] = unNorm.Fingerprint
+
+		/* batch INSERT --------------------------------------------------- */
+		if err := tsBatch.Append(
+			unNorm.Env,
+			unNorm.Temporality,
+			unNorm.MetricName,
+			unNorm.Description,
+			unNorm.Unit,
+			unNorm.Type,
+			unNorm.IsMonotonic,
+			unNorm.Fingerprint,
+			unNorm.UnixMilli,
+			unNorm.Labels,
+			unNorm.Attrs,
+			unNorm.ScopeAttrs,
+			unNorm.ResourceAttrs,
+			unNorm.Normalized,
+		); err != nil {
+			return fmt.Errorf("append TS: %w", err)
+		}
+		rowsInTSBatch++
+		if rowsInTSBatch == maxRowsPerBatch {
+			if err := tsBatch.Send(); err != nil {
+				return fmt.Errorf("flush TS batch: %w", err)
+			}
+			tsBatch, err = newTSBatch()
+			if err != nil {
+				return err
+			}
+			rowsInTSBatch = 0
+		}
+	}
+	if err := rowsTS.Err(); err != nil {
+		return err
+	}
+	/* flush any remainder */
+	if rowsInTSBatch > 0 {
+		if err := tsBatch.Send(); err != nil {
+			return fmt.Errorf("final TS flush: %w", err)
+		}
+	}
+
+	/* ------------------------------------------------------------------
+	   2) READ & WRITE THE SAMPLE ROWS (same batching pattern)
+	   ------------------------------------------------------------------ */
+	if len(fingerprintMap) == 0 {
+		return nil // nothing to do
+	}
+
+	fps := make([]uint64, 0, len(fingerprintMap))
+	for fp := range fingerprintMap {
+		fps = append(fps, fp)
+	}
+
+	querySamples := fmt.Sprintf(`
+		SELECT
+			env, temporality, metric_name, fingerprint,
+			unix_milli, value, flags
+		FROM %s.%s
+		WHERE fingerprint IN (?) and unix_milli >= ? and unix_milli < ?`,
+		signozMetricDBName, signozSampleTableName)
```

Comment:
Make this configurable that there will be n goroutines reading and writing from samples broken down, period/n buckets. This is sequential now. We want to control the speed at which we migrate data. We should be able to migrate the data for entire hours with n routines parallel instead of doing 1 batch after another. I will require higher resources but we will provision them as needed. The goal is to get the migration script run not more than 1 day at max.